File: BOJ 2/11054.cpp
#include <stdio.h>
#include <algorithm>
using namespace std;
int arr[1010];
int dp[1010];
int dp2[1010];
int ans = -2;
int main() {
	int n;
	scanf("%d", &n);
	for (int i = 1; i <= n; i++)
		scanf("%d", &arr[i]);
	for (int i = 1; i <= n; i++)
	{
		dp[i] = 1;
		dp2[i] = 1;
	}
	for (int i = 2; i <= n; i++) {
		for (int j = 1; j<i; j++) {
			if (arr[i]>arr[j])
				dp[i] = max(dp[i], dp[j] + 1);
		}
	}
	for (int i = n-1; i>=1; i--) {
		for (int j = n; j>i; j--) {
			if (arr[i]>arr[j])
				dp2[i] = max(dp2[i], dp2[j] + 1);
		}
	}
	for (int i = 1; i <= n; i++)
		ans = max(ans, dp[i] + dp2[i]-1);
	printf("%d\n", ans);
	return 0;
}

File: BOJ(Java 공부용)/11054.java
import java.io.*;
import java.util.StringTokenizer;

import static java.lang.Integer.max;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0, 0, 1, -1};
    static int[] goy = {1, -1, 0, 0};
    static int counter = 1;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int[] arr = new int[n];
        int[] dp1 = new int[n];
        int[] dp2 = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            dp1[i] = 1;
            dp2[i] = 1;
        }
        int max = 1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                if (arr[j] < arr[i]) {
                    dp1[i] = max(dp1[i], dp1[j] + 1);
                }
            }
        }
        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j >= i; j--) {
                if (arr[j] < arr[i]) {
                    dp2[i] = max(dp2[i], dp2[j] + 1);
                }
            }
        }
        for (int i = 0; i < n; i++) {
            max = max(max, dp1[i] + dp2[i]);
        }
        max--;
        bw.write(max + " ");
        bw.flush();
        bw.close();
    }
}